tabgroups/init.lua

```diff
@@ -0,0 +1,870 @@
+            for i, v in grouptabs(w, gn) do
+                assert(g.tabs[v])
+                state[w].tab_groups.groups[gn].tabs[i] = { tab_hit = g.tabs[v].tab_hit or 0,
+                    atime = g.tabs[v].atime,
+                    mtime = g.tabs[v].mtime,
+                    ctime = g.tabs[v].ctime,
+                }
+                -- we don't store uri and state for tabs in active tg because this info already stored
+                -- by luakit session manager
+                if gn ~= w2groups[w].active then
+                    state[w].tab_groups.groups[gn].tabs[i].uri = v.uri
+                    state[w].tab_groups.groups[gn].tabs[i].session_state = v.session_state
+                end
+                if v == _a then
+                    state[w].tab_groups.groups[gn].active = i
+                end
+            end
+        end
+    end
+    return state
+end)
+
+local function _sort_by_field(field, order, a, b)
+    assert(field)
+    assert(order)
+    assert(a[field])
+    assert(b[field])
+    local va, vb = a[field], b[field]
+    local ret
+    if va <= vb then
+        ret = true
+    else
+        ret = false
+    end
+    if order == 'desc' then
+        ret = not ret
+    end
+    return ret
+end
+
+local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
+    local rows = {{"Group name", "Tab title", "URI", title = true}}
+    local active = current_webview_in_group(w, group_name)
+
+    local _tmp = {}
+    local _gv = w2groups[w].groups[group_name]
+    for i, v in grouptabs(w, group_name) do
+        if not _gv.tabs[v] then
+            _gv.tabs[v] = { tab_hit = 1, atime = os.time(), ctime = os.time() }
+        end
+        table.insert(_tmp, {v = v,
+            hits = _gv.tabs[v].tab_hit,
+            atime = _gv.tabs[v].atime,
+            mtime = _gv.tabs[v].mtime,
+            ctime = _gv.tabs[v].ctime,
+            title = v.title,
+            n = i })
+    end
+    if field and order then
+        table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
+    end
+
+    for i, v in ipairs(_tmp) do
+        local title = '*No title*'
+        if v.title then
+            title = v.v.title
+        end
+        if v.v == active then
+            table.insert(rows, {"<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
+                "<b>"..lousy.util.escape(title).."</b>",
+                lousy.util.escape(v.v.uri),
+                _group = group_name, _tab = v.v})
+        else
+            table.insert(rows, {((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
+                lousy.util.escape(title),
+                lousy.util.escape(v.v.uri),
+                _group = group_name, _tab = v.v})
+        end
+    end
+    return rows
+end
+
+local function _build_tabgroup_menu_grouplist(w, field, order)
+    local rows = {{ "Group name", "Number of tabs", title = true }}
+
+    local _tmp = {}
+    for g, gv in pairs(w2groups[w].groups) do
```

Comment:
this code seems similar to the code in the function above: can it be moved into a helper function?

Reply:
this codes looks similar but dealing with different structures. one with groups another with tabs in group. so i think better leave as is

Reply:
ok no problem then